# <font color=blue>Assignments for "Understanding The Relationship"</font>

To close out this lesson, you're going to do three assignments. For the first assignment, you'll write up a short answer to a question in a Gist file.  For the second two assignments, you'll do your work in Jupyter notebooks, and you should link to those notebooks in the same Gist file.

Please submit a single Gist file containing the answer to first assignment, plus links for second two.

## 1. Interpretion and signficance

Suppose that we would like to know how much families in the US are spending on recreation annually. We estimated the following model:

$$ expenditure = 873 + 0.0012annual\_income + 0.00002annual\_income^2 - 223.57have\_kids $$

*expenditure* is the annual spending on recreation in US dollars, *annual_income* is the annual income in US dollars and *have_kids* is a dummy variable indicating the families with children. Interpret the estimated coefficients. What additional statistics should be given in order to make sure that your interpretations make sense statistically. Write up your answer and save in a Gist. 

Our target variable is expenditure and others variables are annual income,  squared of annual income and having children. When annual income increases and hence squared annuall income increases, expenditure naturally increase. Having kids has negative effect on expenditure. We need to know p-values and t-values, so we need to aapply t-test to understand whether a coefficient is statistically significant.

## 2. Weather model

In this exercise, you'll work with the historical temperature data from the previous lesson. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn import preprocessing
from scipy.stats.mstats import winsorize
import warnings
warnings.filterwarnings('ignore')

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
import pandas as pd
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

- Build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables? 

In [ ]:
weather_df['temp'] = weather_df['Temperature (C)'] - weather_df['Apparent Temperature (C)']
X = weather_df[['Humidity','Wind Speed (km/h)']]
Y = weather_df[['temp']]

In [ ]:
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   temp   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        22:34:50   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* Since  p value is less than 0.005 it is statistically significant as the predicted humidity and wind speed. Both humidity and wind speed increase, the temperature increases by an average of 3.0292 an 0.1193, respectively. *

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for *humidity* and *windspeed* change? Interpret the estimated coefficients.

In [ ]:
weather_df['hwind'] = weather_df['Humidity'] * weather_df['Wind Speed (km/h)']
X1 = weather_df[['Humidity', 'Wind Speed (km/h)', 'hwind']]
Y1 = weather_df[['temp']]

In [ ]:
lrm = linear_model.LinearRegression()
lrm.fit(X1, Y1)

X1 = sm.add_constant(X1)
result1 = sm.OLS(Y1, X1).fit()
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   temp   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        22:40:27   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
hwind                 0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

*  Temperature (C) = -0.0839 - 0.1775Humidity - 0.09053Wind Speed + 0.2971*hwind

Since again p values(0) are less than 0.05 all coefficients in the model are statistically significant. Humidity, wind speed and their interactions affect the temperature in an opposite sign because of all three have negative values.

##  3. House prices model

In this exercise, you'll interpret your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [ ]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

- Run your house prices model again and interpret the results. Which features are statistically significant and which are not?

In [ ]:
house = house_prices_df.drop(['Id','PoolQC','MiscFeature', 'Alley','Fence', 'FireplaceQu', 'LotFrontage'], axis=1)

In [ ]:
house = house.drop(house[(house['GrLivArea']>4000) & (house['SalePrice']<300000)].index)
house = house.drop(house[(house["TotalBsmtSF"]>6000) & (house["SalePrice"]<200000)].index)
house = house.drop(house[(house["TotRmsAbvGrd"]>12) & (house["SalePrice"]<250000)].index)

In [ ]:
house['PoolArea'] = pd.get_dummies(house['PoolArea'],drop_first=True)
house['GarageArea'] = pd.get_dummies(house['GarageArea'],drop_first=True)
house['TotalBsmtSF'] = pd.get_dummies(house['TotalBsmtSF'],drop_first=True)
house['2ndFlrSF'] = pd.get_dummies(house['2ndFlrSF'],drop_first=True)
house['Fireplaces'] = pd.get_dummies(house['Fireplaces'],drop_first=True)

In [ ]:
house['GarageYrBlt'] = house['GarageYrBlt'].fillna(house['YearBuilt'])
house['MasVnrArea'] = house['MasVnrArea'].fillna(house['MasVnrArea'].mean())

In [ ]:
from sklearn.preprocessing import LabelEncoder
categorical_feature_mask = house.dtypes==object
categorical_cols = house.columns[categorical_feature_mask].tolist()
labelencoder = LabelEncoder()
house[categorical_cols] = house[categorical_cols].apply(lambda col: labelencoder.fit_transform(col.astype(str)))

In [ ]:
house['Log_SalePrice']=np.log(house["SalePrice"]) #from previous assignment we know that sale price feature does not have normal distribution so tkae logarithm
house.head()

MSSubClass  MSZoning  LotArea  ...  SaleCondition  SalePrice  Log_SalePrice
0          60         3     8450  ...              4     208500      12.247694
1          20         3     9600  ...              4     181500      12.109011
2          60         3    11250  ...              4     223500      12.317167
3          70         3     9550  ...              0     140000      11.849398
4          60         3    14260  ...              4     250000      12.429216

[5 rows x 75 columns]

In [ ]:
Y2 = house['Log_SalePrice']
X2 = house.drop(['SalePrice','Log_SalePrice'], axis=1)

In [ ]:
lrm = linear_model.LinearRegression()
lrm.fit(X2, Y2)

X2 = sm.add_constant(X2)
result1 = sm.OLS(Y2, X2).fit()
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Log_SalePrice   R-squared:                       0.919
Model:                            OLS   Adj. R-squared:                  0.915
Method:                 Least Squares   F-statistic:                     214.4
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:11:04   Log-Likelihood:                 1097.9
No. Observations:                1457   AIC:                            -2048.
Df Residuals:                    1383   BIC:                            -1657.
Df Model:                          73                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const            18.0682      4.871      3.710      0.000       8.514      27.623
MSSubClass    -7.127e-05      0.000     -0.425      0.671      -0.000       0.000
MSZoning         -0.0166      0.006     -2.995      0.003      -0.028      -0.006
LotArea        2.761e-06   3.91e-07      7.063      0.000    1.99e-06    3.53e-06
Street            0.2076      0.051      4.049      0.000       0.107       0.308
LotShape         -0.0038      0.002     -1.542      0.123      -0.009       0.001
LandContour      -0.0104      0.005     -2.048      0.041      -0.020      -0.000
Utilities        -0.1300      0.122     -1.062      0.289      -0.370       0.110
LotConfig        -0.0022      0.002     -1.092      0.275      -0.006       0.002
LandSlope        -0.0070      0.014     -0.490      0.624      -0.035       0.021
Neighborhood     -0.0003      0.001     -0.604      0.546      -0.001       0.001
Condition1        0.0018      0.004      0.485      0.628      -0.006       0.009
Condition2       -0.0152      0.013     -1.206      0.228      -0.040       0.009
BldgType         -0.0113      0.005     -2.074      0.038      -0.022      -0.001
HouseStyle        0.0012      0.002      0.511      0.609      -0.004       0.006
OverallQual       0.0606      0.004     13.823      0.000       0.052       0.069
OverallCond       0.0426      0.004     11.048      0.000       0.035       0.050
YearBuilt         0.0018      0.000      5.948      0.000       0.001       0.002
YearRemodAdd      0.0006      0.000      2.446      0.015       0.000       0.001
RoofStyle         0.0006      0.004      0.153      0.879      -0.008       0.009
RoofMatl         -0.0039      0.006     -0.710      0.478      -0.015       0.007
Exterior1st      -0.0043      0.002     -2.224      0.026      -0.008      -0.001
Exterior2nd       0.0041      0.002      2.386      0.017       0.001       0.008
MasVnrType        0.0167      0.006      2.976      0.003       0.006       0.028
MasVnrArea     3.052e-05   2.24e-05      1.363      0.173   -1.34e-05    7.45e-05
ExterQual        -0.0123      0.007     -1.699      0.089      -0.027       0.002
ExterCond         0.0112      0.005      2.400      0.017       0.002       0.020
Foundation        0.0137      0.006      2.212      0.027       0.002       0.026
BsmtQual         -0.0113      0.005     -2.170      0.030      -0.022      -0.001
BsmtCond          0.0086      0.005      1.712      0.087      -0.001       0.018
BsmtExposure     -0.0050      0.003     -1.516      0.130      -0.011       0.001
BsmtFinType1     -0.0014      0.002     -0.585      0.559      -0.006       0.003
BsmtFinSF1        0.0002   1.91e-05      8.364      0.000       0.000       0.000
BsmtFinType2      0.0048      0.005      0.950      0.342      -0.005       0.015
BsmtFinSF2        0.0001   3.36e-05      4.259   

- Now, exclude the insignificant features from your model. Did anything change?

In [ ]:
Y2 = house['Log_SalePrice']
X2 = house[['MSSubClass', 'LotArea','OverallCond', 'YearBuilt', 'MasVnrArea', 'BsmtFinSF1', 'TotalBsmtSF',
            '2ndFlrSF', 'GrLivArea', 'BsmtFullBath', 'BedroomAbvGr','TotRmsAbvGrd', 'GarageCars']]

In [ ]:
lrm = linear_model.LinearRegression()
lrm.fit(X2, Y2)

X2 = sm.add_constant(X2)
result1 = sm.OLS(Y2, X2).fit()
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Log_SalePrice   R-squared:                       0.850
Model:                            OLS   Adj. R-squared:                  0.849
Method:                 Least Squares   F-statistic:                     628.9
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:16:19   Log-Likelihood:                 650.73
No. Observations:                1457   AIC:                            -1273.
Df Residuals:                    1443   BIC:                            -1199.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            0.2140      0.352      0.608      0.543      -0.476       0.904
MSSubClass      -0.0009     0.0001     -9.328      0.000      -0.001      -0.001
LotArea       2.527e-06   4.39e-07      5.760      0.000    1.67e-06    3.39e-06
OverallCond      0.0718      0.004     18.092      0.000       0.064       0.080
YearBuilt        0.0054      0.000     30.590      0.000       0.005       0.006
MasVnrArea    5.736e-05    2.6e-05      2.207      0.027    6.37e-06       0.000
BsmtFinSF1       0.0001   1.33e-05      7.822      0.000    7.81e-05       0.000
TotalBsmtSF     -0.0611      0.156     -0.391      0.696      -0.367       0.245
2ndFlrSF        -0.1344      0.156     -0.861      0.389      -0.441       0.172
GrLivArea        0.0005    1.6e-05     28.652      0.000       0.000       0.000
BsmtFullBath     0.0331      0.011      3.103      0.002       0.012       0.054
BedroomAbvGr    -0.0504      0.007     -7.043      0.000      -0.064      -0.036
TotRmsAbvGrd     0.0061      0.005      1.166      0.244      -0.004       0.016
GarageCars       0.0896      0.007     11.981      0.000       0.075       0.104
==============================================================================
Omnibus:                      155.768   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              537.659
Skew:                          -0.499   Prob(JB):                    1.77e-117
Kurtosis:                       5.803   Cond. No.                     1.25e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.25e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- Interpret the statistically significant coefficients by quantifying their relations with the house prices. Which features have more prominent effect on the house prices?


While some coefficients did not change much, some others have changed a lot like number of bathrooms, year of the built, garage area of house.

- Do the results sound reasonable to you? If not, try to explain the potential reasons.

Yes, they are